cmd/sidecars/sidecar_shim.go

```diff
@@ -241,6 +287,35 @@ func main() {
 	hooksInfo.RegisterInfoServer(server, infoServer{Version: version})
 	hooksV1alpha1.RegisterCallbacksServer(server, v1Alpha1Server{})
 	hooksV1alpha2.RegisterCallbacksServer(server, v1Alpha2Server{})
+
+	shutdownChan := make(chan struct{})
+	hooksV1alpha3.RegisterCallbacksServer(server, v1Alpha3Server{done: shutdownChan})
+
+	// Handle signals to properly shutdown process
+	signalStopChan := make(chan os.Signal, 1)
+	signal.Notify(signalStopChan, os.Interrupt,
+		syscall.SIGHUP,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT,
+	)
+
 	log.Log.Infof("shim is now exposing its services on socket %s", socketPath)
-	server.Serve(socket)
+	errChan := make(chan error)
+	go func() {
+		errChan <- server.Serve(socket)
+	}()
+
+	select {
+	case s := <-signalStopChan:
+		log.Log.Infof("Received signal %s", s.String())
+	case err = <-errChan:
+		log.Log.Reason(err).Error("Failed to run grpc server")
+	case <-shutdownChan:
+		log.Log.Info("Exiting")
+	}
+
+	if err == nil {
+		server.Stop()
```

Comment:
This should probably be `server.GracefulStop()`. Since the server now serves in a goroutine, there might be a race with a pending return of a result from the callbacks.

Reply:
Yes, you are right.